// @tanstack/react-query/build/cjs/react-query/src/index.js
/**
 * react-query
 *
 * Copyright (c) TanStack
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../../query-core/build/esm/index.js');
var useQueries = require('./useQueries.js');
var useQuery = require('./useQuery.js');
var QueryClientProvider = require('./QueryClientProvider.js');
var Hydrate = require('./Hydrate.js');
var QueryErrorResetBoundary = require('./QueryErrorResetBoundary.js');
var useIsFetching = require('./useIsFetching.js');
var useIsMutating = require('./useIsMutating.js');
var useMutation = require('./useMutation.js');
var useInfiniteQuery = require('./useInfiniteQuery.js');
var isRestoring = require('./isRestoring.js');



exports.CancelledError = index.CancelledError;
exports.InfiniteQueryObserver = index.InfiniteQueryObserver;
exports.MutationCache = index.MutationCache;
exports.MutationObserver = index.MutationObserver;
exports.QueriesObserver = index.QueriesObserver;
exports.QueryCache = index.QueryCache;
exports.QueryClient = index.QueryClient;
exports.QueryObserver = index.QueryObserver;
exports.dehydrate = index.dehydrate;
exports.focusManager = index.focusManager;
exports.hashQueryKey = index.hashQueryKey;
exports.hydrate = index.hydrate;
exports.isCancelledError = index.isCancelledError;
exports.isError = index.isError;
exports.notifyManager = index.notifyManager;
exports.onlineManager = index.onlineManager;
exports.parseFilterArgs = index.parseFilterArgs;
exports.parseMutationArgs = index.parseMutationArgs;
exports.parseMutationFilterArgs = index.parseMutationFilterArgs;
exports.parseQueryArgs = index.parseQueryArgs;
exports.useQueries = useQueries.useQueries;
exports.useQuery = useQuery.useQuery;
exports.QueryClientProvider = QueryClientProvider.QueryClientProvider;
exports.defaultContext = QueryClientProvider.defaultContext;
exports.useQueryClient = QueryClientProvider.useQueryClient;
exports.Hydrate = Hydrate.Hydrate;
exports.useHydrate = Hydrate.useHydrate;
exports.QueryErrorResetBoundary = QueryErrorResetBoundary.QueryErrorResetBoundary;
exports.useQueryErrorResetBoundary = QueryErrorResetBoundary.useQueryErrorResetBoundary;
exports.useIsFetching = useIsFetching.useIsFetching;
exports.useIsMutating = useIsMutating.useIsMutating;
exports.useMutation = useMutation.useMutation;
exports.useInfiniteQuery = useInfiniteQuery.useInfiniteQuery;
exports.IsRestoringProvider = isRestoring.IsRestoringProvider;
exports.useIsRestoring = isRestoring.useIsRestoring;
