// @moralisweb3/evm-api/lib/resolvers/account/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNFTsForContract = exports.getNFTs = exports.getTransactions = exports.getTokenTransfers = exports.getNFTTransfers = exports.getNativeBalance = exports.getTokenBalances = void 0;
var getTokenBalances_1 = require("./getTokenBalances");
Object.defineProperty(exports, "getTokenBalances", { enumerable: true, get: function () { return getTokenBalances_1.getTokenBalances; } });
var getNativeBalance_1 = require("./getNativeBalance");
Object.defineProperty(exports, "getNativeBalance", { enumerable: true, get: function () { return getNativeBalance_1.getNativeBalance; } });
var getNFTTransfers_1 = require("./getNFTTransfers");
Object.defineProperty(exports, "getNFTTransfers", { enumerable: true, get: function () { return getNFTTransfers_1.getNFTTransfers; } });
var getTokenTransfers_1 = require("./getTokenTransfers");
Object.defineProperty(exports, "getTokenTransfers", { enumerable: true, get: function () { return getTokenTransfers_1.getTokenTransfers; } });
var getTransactions_1 = require("./getTransactions");
Object.defineProperty(exports, "getTransactions", { enumerable: true, get: function () { return getTransactions_1.getTransactions; } });
var getNFTs_1 = require("./getNFTs");
Object.defineProperty(exports, "getNFTs", { enumerable: true, get: function () { return getNFTs_1.getNFTs; } });
var getNFTsForContract_1 = require("./getNFTsForContract");
Object.defineProperty(exports, "getNFTsForContract", { enumerable: true, get: function () { return getNFTsForContract_1.getNFTsForContract; } });
