// @moralisweb3/auth/lib/resolvers/evmRequestChallenge.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeChallenge = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var MoralisAuth_1 = require("../MoralisAuth");
var method = 'post';
var bodyParams = [
    'domain',
    'chainId',
    'address',
    'statement',
    'uri',
    'expirationTime',
    'notBefore',
    'resources',
    'timeout',
];
var apiToResult = function (apiData, params) {
    var data = (0, core_1.toCamelCase)(apiData);
    return __assign({}, data);
};
exports.initializeChallenge = (0, api_utils_1.createEndpointFactory)(function () {
    return (0, api_utils_1.createEndpoint)({
        name: 'Request Challenge (EVM)',
        getUrl: function (params) { return "".concat(MoralisAuth_1.BASE_URL, "/challenge/request/evm"); },
        apiToResult: apiToResult,
        resultToJson: function (data) { return (__assign({}, data)); },
        parseParams: function (params) { return params; },
        method: method,
        bodyParams: bodyParams,
    });
});
