// @moralisweb3/api-utils/lib/resolvers/ApiPaginatedResultAdapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiPaginatedResultAdapter = void 0;
var core_1 = require("@moralisweb3/core");
var ApiResultAdapter_1 = require("./ApiResultAdapter");
var ApiPaginatedResultAdapter = /** @class */ (function (_super) {
    __extends(ApiPaginatedResultAdapter, _super);
    function ApiPaginatedResultAdapter(data, adapter, jsonAdapter, params, nextCall) {
        var _this = _super.call(this, data, adapter, jsonAdapter, params) || this;
        _this.next = function () {
            if (!_this._nextCall) {
                throw new core_1.MoralisApiError({
                    code: core_1.ApiErrorCode.PAGE_LIMIT_EXCEEDED,
                    message: 'Cannot call .next(). Page limit exceeded.',
                });
            }
            return _this._nextCall();
        };
        _this._nextCall = nextCall;
        return _this;
    }
    Object.defineProperty(ApiPaginatedResultAdapter.prototype, "pagination", {
        get: function () {
            return {
                total: this._data.total,
                page: this._data.page,
                pageSize: this._data.page_size,
                cursor: this._data.cursor,
            };
        },
        enumerable: false,
        configurable: true
    });
    return ApiPaginatedResultAdapter;
}(ApiResultAdapter_1.ApiResultAdapter));
exports.ApiPaginatedResultAdapter = ApiPaginatedResultAdapter;
